### PR TITLE
Improve unit test coverage

### PR DESCRIPTION
**Pull request by nilasini**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java [hunk 1]
```diff
@@ -356,23 +439,39 @@ public void processAuthenticationResponseTestcaseWithuserStoreException() throws
...
     }
 
-    @Test(dataProvider = "statusProvider")
-    public void initiateAuthenticationRequestTestcaseWithUnknownErrorCode(boolean statusProvider) throws
+    @DataProvider(name = "enableStatusProvider")
+    public Object[][] getEnabledOption() {
+
+        return new String[][]{
+                {"true", "false"},
+                {"false", null},
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java]: what about other combinations?
like null, null

Comment (nilasini) [reply components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java]: First argument is used for checking isEnable condition. Even though it is String it is used as boolean so in that case there is no possibility for null.

File: components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java [hunk 2]
```diff
@@ -303,12 +321,79 @@ public Object answer(InvocationOnMock invocation) throws Throwable {
...
+            }
+        }).when(mockLog).debug(anyString());
+
         basicAuthenticator.processAuthenticationResponse(mockRequest, mockResponse, mockAuthnCtxt);
 
-        assertEquals(authenticatedUser.getAuthenticatedSubjectIdentifier(), dummyUserName);
+        if (userNameValue
+                != null && userNameValue.trim().length() > 0) {
+            assertEquals(authenticatedUser.getAuthenticatedSubjectIdentifier(), userNameValue +
+                    CarbonConstants.DOMAIN_SEPARATOR + dummyDomainName + "@" + "dummyTenantDomain");
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java]: is this correct?

dummyDomainName + CarbonConstants.DOMAIN_SEPARATOR + userNameValue + "@" + "dummyTenantDomain" ?

File: components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java [hunk 3]
```diff
@@ -303,12 +321,79 @@ public Object answer(InvocationOnMock invocation) throws Throwable {
...
+            assertEquals(authenticatedUser.getAuthenticatedSubjectIdentifier(), dummyUserName);
+        }
         if (Boolean.valueOf(chkRemember)) {
             assertEquals(isrememberMe, (Boolean) true);
         }
+        if (Boolean.parseBoolean(debugEnabled) && userNameUri != null && userNameUri.trim().length() > 0) {
+            if (Boolean.valueOf(multipleAttributeEnable) && userNameValue != null &&
+                    userNameValue.trim().length() > 0) {
+                assertEquals(debugMsg, "UserNameAttribute is found for user. Value is :  " + userNameValue +
+                        CarbonConstants.DOMAIN_SEPARATOR + dummyDomainName + "@" + "dummyTenantDomain");
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java]: is this correct?

dummyDomainName + CarbonConstants.DOMAIN_SEPARATOR + userNameValue + "@" + "dummyTenantDomain" ?

File: components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java [hunk 4]
```diff
@@ -226,64 +241,67 @@ public void processAuthenticationResponseTestCaseForException(Object realm, Obje
...
-        mockUserStoreManager = mock(UserStoreManager.class);
-        when(BasicAuthenticatorServiceComponent.getRealmService().getTenantUserRealm(dummyTenantId)).thenReturn(mockRealm);
-        when(mockRealm.getUserStoreManager()).thenReturn(mockUserStoreManager);
+        mockRealmConfiguration = mock(RealmConfiguration.class);
 
-        mockStatic(MultitenantUtils.class);
-        when(MultitenantUtils.getTenantAwareUsername(dummyUserName)).thenReturn(dummyUserName);
-        when(mockUserStoreManager.authenticate(
-                MultitenantUtils.getTenantAwareUsername(dummyUserName), dummyUserName)).thenReturn(true);
+        if (domainName != null && domainName.trim().length() > 0) {
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java]: Can use StringUtils.isNotBlank

File: components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java [hunk 5]
```diff
@@ -303,12 +321,79 @@ public Object answer(InvocationOnMock invocation) throws Throwable {
...
+                debugMsg = (String) invocation.getArguments()[0];
+                return null;
+            }
+        }).when(mockLog).debug(anyString());
+
         basicAuthenticator.processAuthenticationResponse(mockRequest, mockResponse, mockAuthnCtxt);
 
-        assertEquals(authenticatedUser.getAuthenticatedSubjectIdentifier(), dummyUserName);
+        if (userNameValue
+                != null && userNameValue.trim().length() > 0) {
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java]: Can use StringUtils.isNotBlank

File: components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java [hunk 6]
```diff
@@ -303,12 +321,79 @@ public Object answer(InvocationOnMock invocation) throws Throwable {
...
+            assertEquals(authenticatedUser.getAuthenticatedSubjectIdentifier(), dummyDomainName +
+                    CarbonConstants.DOMAIN_SEPARATOR + userNameValue + "@" + "dummyTenantDomain");
+        } else {
+            assertEquals(authenticatedUser.getAuthenticatedSubjectIdentifier(), dummyUserName);
+        }
         if (Boolean.valueOf(chkRemember)) {
             assertEquals(isrememberMe, (Boolean) true);
         }
+        if (Boolean.parseBoolean(debugEnabled) && userNameUri != null && userNameUri.trim().length() > 0) {
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java]: Can use StringUtils.isNotBlank, check and correct at the other similar places also

File: components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java [hunk 7]
```diff
@@ -356,23 +440,40 @@ public void processAuthenticationResponseTestcaseWithuserStoreException() throws
...
+    public void initiateAuthenticationRequestTestcaseWithUnknownErrorCode(String statusProvider, String
+            showAuthFailureReason, String cntxpropUserTenantDomainMismatch) throws
             AuthenticationFailedException, IOException, NoSuchFieldException, IllegalAccessException {
 
         mockAuthnCtxt = mock(AuthenticationContext.class);
         mockRequest = mock(HttpServletRequest.class);
         mockResponse = mock(HttpServletResponse.class);
 
+        AuthenticatorConfig authenticatorConfig = new AuthenticatorConfig();
+        HashMap<String, String> paramMap = new HashMap<>();
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java]: Use Map interface when assigning variable

File: components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java (hunk 6 above)
Comment (nilasini) [reply components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java]: Fixed with https://github.com/wso2-extensions/identity-local-auth-basicauth/pull/43

File: components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java (hunk 5 above)
Comment (nilasini) [reply components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java]: Fixed with https://github.com/wso2-extensions/identity-local-auth-basicauth/pull/43

File: components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java (hunk 4 above)
Comment (nilasini) [reply components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java]: Fixed with https://github.com/wso2-extensions/identity-local-auth-basicauth/pull/43

File: components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java (hunk 7 above)
Comment (nilasini) [reply components/org.wso2.carbon.identity.application.authenticator.basicauth/src/test/java/org/wso2/carbon/identity/application/authenticator/basicauth/BasicAuthenticatorTestCase.java]: Fixed with https://github.com/wso2-extensions/identity-local-auth-basicauth/pull/43

**Pull request merged**